src/main/java/com/google/api/gax/grpc/OperationFuture.java

```diff
@@ -0,0 +1,372 @@
+                responseClass,
+                pollingInterval,
+                waiter,
+                asyncCompletionLatch));
+    OperationFuture<ResponseT> operationFuture =
+        new OperationFuture<>(
+            startOperationFuture, finalResultFuture, dataGetterFuture, asyncCompletionLatch);
+    return operationFuture;
+  }
+
+  private OperationFuture(
+      ListenableFuture<Operation> startOperationFuture,
+      SettableFuture<ResponseT> finalResultFuture,
+      Future<ResponseT> dataGetterFuture,
+      CountDownLatch asyncCompletionLatch) {
+    this.startOperationFuture = startOperationFuture;
+    this.finalResultFuture = finalResultFuture;
+    this.dataGetterFuture = dataGetterFuture;
+    this.asyncCompletionLatch = asyncCompletionLatch;
+  }
+
+  private static class DataGetterRunnable<ResponseT extends Message>
+      implements Callable<ResponseT> {
+    private final ListenableFuture<Operation> startOperationFuture;
+    private final SettableFuture<ResponseT> finalResultFuture;
+    private final OperationsApi operationsApi;
+    private final Class<ResponseT> responseClass;
+    private final Duration pollingInterval;
+    private final Waiter waiter;
+    private final CountDownLatch asyncCompletionLatch;
+
+    public DataGetterRunnable(
+        ListenableFuture<Operation> startOperationFuture,
+        SettableFuture<ResponseT> finalResultFuture,
+        OperationsApi operationsApi,
+        Class<ResponseT> responseClass,
+        Duration pollingInterval,
+        Waiter waiter,
+        CountDownLatch asyncCompletionLatch) {
+      this.startOperationFuture = startOperationFuture;
+      this.finalResultFuture = finalResultFuture;
+      this.operationsApi = operationsApi;
+      this.responseClass = responseClass;
+      this.pollingInterval = pollingInterval;
+      this.waiter = waiter;
+      this.asyncCompletionLatch = asyncCompletionLatch;
+    }
+
+    @Override
+    public ResponseT call() {
+      try {
+        callImpl();
+      } finally {
+        asyncCompletionLatch.countDown();
+      }
+      return null;
+    }
+
+    public void callImpl() {
+      Operation firstOperation = null;
+      try {
+        firstOperation = startOperationFuture.get();
+        if (firstOperation.getDone()) {
+          setResultFromOperation(finalResultFuture, firstOperation, responseClass);
+          return;
+        }
+        do {
+          // TODO: switch implementation from polling to scheduled execution
+          waiter.wait(pollingInterval);
+          Operation latestOperation = operationsApi.getOperation(firstOperation.getName());
+          if (latestOperation.getDone()) {
+            if (isCancelled(latestOperation)) {
+              finalResultFuture.cancel(true);
+            } else {
+              setResultFromOperation(finalResultFuture, latestOperation, responseClass);
+            }
+            return;
+          }
+        } while (true);
+      } catch (InterruptedException e) {
```

Comment:
It occurs to me that as it is written, the code will either leave the polling loop running (`mayInterruptIfRunning = false`), or it will cancel the polling loop and cancel the remote operation (`mayInterruptIfRunning = true`). Instead, if `mayInterruptIfRunning = false`, it should probably still cancel the polling loop but let the remote operation keep running. That would allow users to cancel the future (unhook essentially) without cancelling the remote operation. Since there is some multi-threaded coordination involved in that change, I'd like to roll it into the work to convert the polling loop to chained events (I have now filed https://github.com/googleapis/gax-java/issues/147 to track it).